[#50] 내 동네 등록 Controller 구현

Srltas opened the pull request:
## 📌 Summary
내 동네 등록 기능에 대한 Controller를 구현했습니다.

## 📝 Description
**내 동네 등록 Controller 추가**
- `/users/neighborhoods` POST 요청을 보내면 내 동네 등록 기능을 수행하도록 만들었습니다.
- 등록에 성공하면 200 응답을 보내도록 만들었습니다.

**필터 설정 수정**
- 내 동네 등록도 로그인 된 사용자만 접근할 수 있도록 필터 설정에 내 동네 등록 URL도 추가했습니다.

## ✅ Checklist
- [x] 새로운 기능이나 수정된 기능에 대해 충분한 테스트를 작성했습니다.
- [x] 코딩 스타일 가이드를 준수했습니다.
- [x] 문서(주석, README 등)가 필요하다면 업데이트했습니다.

Srltas committed:
feat: 내 동네 등록 Controller 구현

Srltas committed:
test: 내 동네 등록 Controller 테스트 코드 작성

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/UserNeighborhoodController.java:
```diff
@@ -0,0 +1,44 @@
+		summary = "내 동네 등록",
+		description = "사용자가 특정 동네를 자신의 동네로 등록합니다."
+	)
+	@ApiResponse(responseCode = "200", description = "내 동네 등록 성공")
```
PR Comment 에서는 201 을 내려 준다고 했는데 여기엔 200 으로 되어 있네요.
어떤게 맞는건가요??

Srltas replied:
201 코드로 작성해야 해야 했는데 실수를 했습니다..
근데 상태 코드에 대해 한번 더 고민을 해봤습니다.

처음에는 "내 동네 등록" 기능을 새로운 리소스를 생성하는 동작으로 생각해 201 코드를 반환 했습니다. 하지만, 조금 더 고민을 해보니 이 기능은 이미 존재하는 동네를 사용자 계정에 연결하는 기능입니다.

새로운 리소스를 생성하는 것이 아니라, 기존 리소스를 설정하는 작업에 가깝기 때문에 200 코드를 반환하는 것이 더 적절하다고 생각합니다.

혹시 멘토님의 생각은 어떠신가요?

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/UserNeighborhoodController.java:
```diff
@@ -0,0 +1,44 @@
+	@PostMapping(USER_NEIGHBORHOOD_REGISTRATION)
+	public ResponseEntity<Void> addUserNeighborhood(
+		@RequestBody @Valid AddUserNeighborhoodRequest request, HttpSession session) {
+		UserSessionDTO userSession = (UserSessionDTO)session.getAttribute(USER_SESSION);
```
이 함 수가 굉장히 많이 쓰일 것 같은데요...
함수로 따로 빼는건 어때요?

Srltas replied:
네! 함수로 사용하는 것이 더 좋을거 같습니다 👍

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/UserNeighborhoodController.java:
```diff
@@ -0,0 +1,44 @@
+		UserSessionDTO userSession = (UserSessionDTO)session.getAttribute(USER_SESSION);
+		addUserNeighborhood.addNeighborhood(
+			new AddUserNeighborhoodCommand(userSession.userId(), request.neighborhoodId()));
+		return new ResponseEntity<>(CREATED);
```
ResponseEntity 를 new 말고 어떤걸 사용해서 만들 수 있을까요???

Srltas replied:
ResponseEntity의 `created` 메소드를 사용해 만들 수 있을거 같습니다.

만약, 상태 코드를 200으로 변경하면 `ok` 메소드를 사용해 수정하겠습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/web/common/UrlConstants.java:
```diff
@@ -5,4 +5,5 @@
 @UtilityClass
 public class UrlConstants {
 	public final String NEIGHBORHOOD_VERIFICATION = "/neighborhood/verification";
+	public final String USER_NEIGHBORHOOD_REGISTRATION = "/user-neighborhood";
```
url 이 와닿지 않긴 해요..
어떤 논리로 url 을 작성 하신거에요???

Srltas replied:
처음에는 `userneighborhood`로 작성을 했지만, user와 neighborhood 단어의 구분이 힘들어 `userNeighborhood`로 변경했습니다. 하지만, 이것 또한 URL이 아니라 변수명을 넣은 느낌이라 어색하다고 생각했습니다.
대안으로 `-`을 추가해 user와 neighborhood를 구분해 조금 더 명확하게 단어를 작성하고자 했습니다.

더 고민을 해보니 이 기능 또한 사용자 애그리거트에 속하는 기능으로 URL을 `/user/neighborhood`로 구분해서 작성하는 방법이 애그리거트 관계를 표현하기에 적절한 URL인거 같습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/web/dto/AddUserNeighborhoodRequest.java:
```diff
@@ -0,0 +1,11 @@
+public record AddUserNeighborhoodRequest(
+	@NotNull
+	@PositiveOrZero(message = "동네 ID는 0 이상의 값이어야 합니다.")
+	int neighborhoodId
```
int 타입인데 null 일수가 있나요????
만약 실수로 해당 id 를 입력하지 않으면 전부 0  으로 초기화 될 것같아요.

Srltas replied:
Integer로 타입을 변경해 설정한 조건들을 만족할 수 있도록 수정하겠습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/config/WebConfig.java:
```diff
@@ -21,7 +21,7 @@ public class WebConfig {
 		FilterRegistrationBean<AuthenticationFilter> registrationBean = new FilterRegistrationBean<>();
 		registrationBean.setFilter(new AuthenticationFilter(sessionStorage));
-		registrationBean.addUrlPatterns(NEIGHBORHOOD_VERIFICATION);
+		registrationBean.addUrlPatterns(NEIGHBORHOOD_VERIFICATION, USER_NEIGHBORHOOD_REGISTRATION);
```
이렇게 점점 늘어 날 것 같은데 얘들은 따로 list 로 관리하는건 어때요??

Srltas replied:
List로 관리하는 것이 더 효율적일거 같습니다 👍

Srltas committed:
refactor: 세션에서 사용자 정보를 가져오는 로직을 SessionUtils로 분리

Srltas committed:
refactor: 필터링 대상 URL들을 List로 사용하도록 수정

Srltas committed:
refactor: 동네 ID 참조 타입으로 변경

- 동네 등록 기능에 사용하는 DTO로 동네 ID가 NULL이면 안됩니다
- 동네 ID는 0 이상의 값으로 이루어져있습니다

Srltas committed:
refactor: 내 동네 등록 성공 시 200 반환 및 URL 수정

- 내 동네 등록 성공 시 Http 상태 코드 200을 반환합니다
- 내 동네 등록은 user 애그리거트 안에 속하므로 URL도 users/neighborhoods로 변경합니다

f-lab-bradley reviewed (approved):
수고하셨습니다~~
